
RunATAC
---
## An R package for processing and plotting ATAC-seq data

This package is currently is it's infantcy. Don't expect any real functionality for a while...

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Plot ATAC-seq insert size histogram
```{r}
library(RunATAC)
bam <- system.file("extdata", "chrs_I_M.bam", package = "RunATAC")
#frags <- read_atac_frags(bam_file = chr19)
#plot_insert_size(frags)
```

### Generate a Tn5 insertion site bigwig file
```{r}

```

### Calculate proportion of reads mapped to mitochondrial genome
```{r}

```

### Calculate the fraction of reads in peaks (FrIP)
```{r}

```

### Generate a read counts table for peaks
```{r}

```

### Plotting ATAC-seq footprints

In this example, we will plot the Tn5 insertion signal around CTCF motifs in ATAC-seq peaks using a positional weight matrix (PWM).

Get the REB1 motif from the JASPAR database.
```{r, message=FALSE, comment=FALSE}
library(JASPAR2016)
library(TFBSTools)

source("https://bioconductor.org/biocLite.R")
biocLite("BSgenome.Scerevisiae.UCSC.sacCer3")
library(BSgenome.Scerevisiae.UCSC.sacCer3)

reb1 <- getMatrixByName(JASPAR2016, name=c("REB1"))

```

Load the bed file of ATAC-seq peaks.
```{r}
peaks <- read_bed("~/polo_iPSC/ATACseq/processed_data/atac_peaks/atac_all_peaks_union.bed")
```

Read in the BAM and get insertion signal for regions.
```{r}
tn <- read_atac_bam_tn(bam_file = "~/Desktop/test.bam", yieldSize = 1e6, which=rgns)
```

Get the positions of motif in peaks.
```{r}
gr_motif <- motif_gr(gr = peaks, pwm = pwm, genome = Mmusculus)
```

Get the distances from motif.
```{r}
dists <- atac_motif_dist(tn_signal_gr = tn, motif_pos_gr = gr_motif, range = 150)
```

Plot the footprint.
```{r}
plot_scaled_footprint(dists, window = 300, col='red', lwd=2, main="CTCF")
```

Generate V-plot for CTCF motif
```{r}

```


