
RunATAC
---
## An R package for processing and plotting ATAC-seq data

This package is currently is it's infantcy. Don't expect any real functionality for a while...

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


### Install and load the package
```{r}
devtools::install_github("SamBuckberry/RunATAC")
library(RunATAC)
```

### Plot ATAC-seq insert size histogram

First, load a BAM file. Here we will use a BAM file included in the RunATAC package. This BAM is from ATAC-seq on Saccharomyces cerevisiae and includes data only from chrIV. 
```{r}
bam_fl <- system.file("extdata", "chrIV.bam", package = "RunATAC")
frags <- read_atac_frags(bam_file = bam_fl)
```

Plot the insert size distribution
```{r, fig.height=3, fig.width=5}
plot_insert_size(frags)
```


### Generate a Tn5 insertion site bigwig file
```{r}

```

### Calculate proportion of reads mapped to mitochondrial genome
```{r}

```

### Calculate the fraction of Tn5 insetions in peaks (FrIP)

Get the Tn5 insertion points from BAM file into GRanges object
```{r}
ins <- read_atac_bam_tn(bam_fl)
ins
```

Load the peak data in the MACS2 .narrowPeak format
```{r}
peak_fl <- system.file("extdata", "chrIV_peaks.narrowPeak", package = "RunATAC")
peaks <- read_bed(peak_fl)
peaks
```

Calculate the fraction of reads in peaks
```{r}
hits <- overlapsAny(query = ins, subject = peaks)
FrIP <- sum(hits) / length(ins)
FrIP
```


### Generate a read counts table for peaks
```{r}

```

### Plotting ATAC-seq footprints

In this example, we will plot the Tn5 insertion signal around CTCF motifs in ATAC-seq peaks using a positional weight matrix (PWM).

Get the REB1 motif from the JASPAR database.
```{r, message=FALSE, comment=FALSE}
#source("https://bioconductor.org/biocLite.R")
#biocLite("BSgenome.Scerevisiae.UCSC.sacCer3")
library(BSgenome.Scerevisiae.UCSC.sacCer3)
library(JASPAR2016)
library(TFBSTools)
reb1 <- getMatrixByName(JASPAR2016, name=c("REB1"))
```

Load the bed file of ATAC-seq peaks.
```{r}
peaks <- read_bed("~/polo_iPSC/ATACseq/processed_data/atac_peaks/atac_all_peaks_union.bed")
```

Read in the BAM and get insertion signal for regions.
```{r}
tn <- read_atac_bam_tn(bam_file = "~/Desktop/test.bam", yieldSize = 1e6, which=rgns)
```

Get the positions of motif in peaks.
```{r}
gr_motif <- motif_gr(gr = peaks, pwm = pwm, genome = Mmusculus)
```

Get the distances from motif.
```{r}
dists <- atac_motif_dist(tn_signal_gr = tn, motif_pos_gr = gr_motif, range = 150)
```

Plot the footprint.
```{r}
plot_scaled_footprint(dists, window = 300, col='red', lwd=2, main="CTCF")
```

Generate V-plot for CTCF motif
```{r}

```


