---
    title: "RunATAC"
    author: "Sam Buckberry"
    date: "`r format(Sys.Date())`"
    output: github_document
---

RunATAC
---
## An R package for processing and plotting ATAC-seq data

This package is currently is it's infantcy. Don't expect any real functionality for a while...

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = FALSE, results = "hide",
                      comment = FALSE, warning = FALSE, message=FALSE)
```

### Install the RunATAC package and dependencies
```{r, eval=FALSE}
library(devtools)
devtools::install_github("SamBuckberry/RunATAC")

install.packages('data.table')
install.packages('magrittr')

## try http:// if https:// URLs are not supported
source("https://bioconductor.org/biocLite.R")
biocLite(c("Biostrings", "motifRG", "GenomicRanges", "IRanges", "GenomicAlignments", "rtracklayer", "Rsamtools"))
```

### Plot ATAC-seq insert size histogram

First, load a BAM file. Here we will use a BAM file included in the RunATAC package. This BAM is from ATAC-seq on Saccharomyces cerevisiae and includes data only from chrIV. 
```{r}
library(RunATAC)
bam_fl <- system.file("extdata", "chrIV.bam", package = "RunATAC")
frags <- read_atac_frags(bam_file = bam_fl)
```

Plot the insert size distribution. 
```{r, fig.height=3, fig.width=5}
plot_insert_size(frags, xlim=c(0,600))
```

### Generate a Tn5 insertion site and nucleosome bigwig files
```{r, eval=FALSE}
bam_to_insertions_bw(bam_file = bam_fl, file = "inst/extdata/chrIV.ins.bigwig")
bam_to_centres_bw(bam_file = bam_fl, file = "inst/extdata/chrIV.nuc.bigwig")
```


### Calculate the fraction of Tn5 insetions in peaks (FrIP)

Get the Tn5 insertion points from BAM file into GRanges object
```{r}
ins <- read_atac_insertions(bam_fl)
ins
```

Load the peak data in the MACS2 .narrowPeak format
```{r}
peak_fl <- system.file("extdata", "chrIV_peaks.narrowPeak", package = "RunATAC")
peaks <- read_bed(peak_fl)
peaks
```

Calculate the fraction of reads in peaks
```{r}
calc_frip(signal = ins, peaks = peaks)
```


### Generate a read counts table for peaks
```{r}
library(GenomicAlignments)
library(stringr)
olap_counts <- summarizeOverlaps(features = peaks, reads = ins)
olap_counts <- assays(olap_counts)$counts
rownames(olap_counts) <- str_c(seqnames(peaks), start(peaks), sep = ":") %>% 
        str_c(end(peaks), sep = "-")

```

### Calculate di-nucleotide insertion frequency
```{r}

library(BSgenome.Scerevisiae.UCSC.sacCer3)

dinuc_freq <- calc_dinuc_freq(ins_gr = ins, genome = Scerevisiae)
barplot(dinuc_freq$percentage, names=dinuc_freq$dinucleotide)
```


### Generate tn5 insertion PWM
```{r}
ins_pwm <- calc_ins_pwm(ins_gr = ins, genome = Scerevisiae, seq_width = 40)

library(reshape2)
library(ggplot2)

dat <- melt(ins_pwm)
gg <- ggplot(dat, aes(x = Var2, y = value, group=Var1, colour=Var1)) +
        geom_line()
gg

```


### Calculate genome-wide Tn5 insertion bias bigwig
```{r}
regions_gr <- peaks
genome <- Scerevisiae

calc_ins_bias <- function(ins_pwm, regions_gr, genome){
        
        # Pad the end of the ranges with sequence for PWM scores 
        regions_pad <- regions_gr
        strand(regions_pad) <- "*"
        end(regions_pad) <- end(regions_pad) + ncol(ins_pwm)
        
        # Get the sequences for the regions of interest
        region_seq <-  BSgenome::getSeq(x = genome, regions_pad)
        
        
        
        calc_seq_pwm_score <- function(x){
                
                dna <- region_seq[[x]]
                pwm_score <- Biostrings::PWMscoreStartingAt(pwm = ins_pwm,
                                                            subject = dna,
                                                            starting.at = 1:(length(dna)-ncol(ins_pwm)))
        }
        
        scores <- lapply(1:length(region_seq), FUN = calc_seq_pwm_score)
        
        return(scores)
}



```


### Plotting ATAC-seq footprints

In this example, we will plot the Tn5 insertion signal around CTCF motifs in ATAC-seq peaks using a positional weight matrix (PWM).

Get the REB1 motif from the JASPAR database.
```{r, message=FALSE, comment=FALSE}
#source("https://bioconductor.org/biocLite.R")
#biocLite(c("BSgenome.Scerevisiae.UCSC.sacCer3", "JASPAR2016", "TFBSTools"))
library(JASPAR2016)
library(TFBSTools)

reb1 <- getMatrixByName(JASPAR2016, name=c("REB1"))
reb1 <- reb1@profileMatrix
```

Get the positions of motif in peaks.
```{r}
motif_pos <- motif_gr(gr = peaks, pwm = reb1, genome = Scerevisiae)
motif_pos
```

Get the insertion data for the motif regions
```{r, fig.height=4, fig.width=4.5, eval=FALSE}
ins_bw <- system.file("extdata", "chrIV.ins.bigwig", package = "RunATAC")
dat <- range_summary(bigwig = ins_bw, gr = motif_pos)

library(reshape2)
library(dplyr)
library(ggplot2)
mdat <- melt(dat)
fp <- mdat %>% group_by(variable) %>%
                        summarise(yy=mean(value))
fp$variable <- as.character(fp$variable) %>% as.numeric()


gg <- ggplot(fp, aes(x = variable, y = yy, fill = 'blue', title="")) +
                #stat_smooth(span = 10) +
                #geom_area(aes(fill = 'blue', stat = "bin")) +
                geom_line(size=0.3) +
                xlab("Distance from motif") +
                ylab("Tn5 insertions") +
                ggtitle(title) +
                theme_bw() +
                theme(plot.background = element_blank(),
                      panel.border = element_blank(),
                      strip.background = element_rect(size = 0),
                      panel.grid.minor = element_blank(),
                      axis.line = element_line(),
                      axis.text = element_text(size=8))
gg
```

Generate V-plot for motif regions
```{r, fig.height=4, fig.width=4.5}
v_dat <- calc_v(frags_gr = frags, motif_pos_gr = motif_pos,
                flank = 200, max_frag = 400)

plot_v(df = v_dat)
```

